### PR TITLE
add fixes

### PR DESCRIPTION
**Pull request by lue-bird**
The rule now suggests fixes:
#### Destructure the variable pattern (for example the argument)
```elm
pointless : () -> Bool
pointless unit =
    case unit of
        () ->
            True

unpack : Opaque -> Int
unpack o =
    case o of
        Opaque i ->
            i
```
would be fixed like this:
```elm
pointless : () -> Bool
pointless () =
    True

unpack : Opaque -> Int
unpack (Opaque i) =
    i
```

This can't be used if
  - the case-block expression isn't just 1 value, for example

      ```elm
      case functionCall o of
      ```

  - the variable wasn't defined in a pattern (e.g. an argument), for example

      ```elm
      import Somewhere exposing (imported)
      case imported of

      topLevel : Opaque
      case topLevel of

      let
          withAnnotation : Opaque
      in
      case withAnnotation of
      ```

  - the variable is a record field or after `as`, for example

      ```elm
      unpack ((Opaque i) as o) =
          case o of

      unpack { o } =
          case o of
      ```
  - the variable is also used somewhere else, for example

      ```elm
      let
          something =
              alsoUse o
      in
      case o of
      ```
In those situations,
#### Replace the case with a destructuring let
```elm
unpacked : Int
unpacked =
    case topLevel of
        Opaque i -> i
```
when fixed:
```elm
unpacked : Int
unpacked =
    let
        (Opaque i) =
            topLevel
    in
    i
```

### Why
- easier refactoring

### Why not
- the rule's code is more difficult to understand
- there often isn't a definite way to fix
```elm
unpack : Opaque -> Int
unpack o =
    let
        something =
            alsoUse o
    in
    case o of
        Opaque i -> i
```
could for example also be fixed like that:
```elm
unpack : Opaque -> Int
unpack (o as (Opaque i)) =
    let
        something =
            alsoUse o
    in
    i
```
or like how it's done currently (which I definitely prefer but others may not).
### final words
Don't merge this now. The change should be _discussed_ first.

### CODE REVIEW

**Comment by SiriusStarr** [comment]
Thanks for the (extraordinary) amount of work on this.  :-)  I'm quite busy at the moment so forgive me if I'm slow at working on this.

RE: there not being a definite way to fix, I assume you mean fixing:

```elm
unpack : Opaque -> Int
unpack o =
    let
        something =
            alsoUse o
    in
    case o of
        Opaque i -> i
```
as
```elm
unpack : Opaque -> Int
unpack o =
    let
        something =
            alsoUse o

        (Opaque i) =
            o
    in
    i
```
versus
```elm
unpack : Opaque -> Int
unpack ((Opaque i) as o) =
    let
        something =
            alsoUse o
    in
    i
```
?

I personally prefer the latter style (using `as`), so perhaps we should make the rule take an argument for preferring fixing with `let` annotations vs `as`?  That way one could adopt either style for a codebase, since it's purely stylistic.

**Comment by lue-bird** [comment]
Using `as` (and joined lets) has the same problem as directly destructuring the variable pattern:
They have a different scope than the original case
```elm
unpack o =
    let
        something i = use i
    in
    case o of
        Opaque i -> i
```
fixed as
```elm
unpack o =
    let
        something i = use i

        (Opaque i) = o
    in
    i
```
Expands the scope of `i` to the expression after `in` **plus** all let declarations:
```elm
let
    something i = use i
    --        ^ already used!
    
    (Opaque i) = o
```
The scope is even wider with `as` & directly destructured variable patterns! It's same as `o`'s scope.

→ `as`, joined lets and directly destructured variable pattern can only be used if there is **no variable pattern with the same name in the new scope**.

- `as`, joined lets and directly destructured variable patterns are shorter
- separate lets use the same (minimal) scope as the original case

In the end, like you suggested, this is a trade-off where the user should decide!

```elm
type DestructuringMethod
    = -- as and direct destructuring only for arguments that only the case..of could access
      -- no destructuring in existing let
      DestructureInMinimalScope { direct : Bool, as_ : Bool }
      -- joined lets, as and direct destructuring possible
    | DestructureInAnyScope DestructuringMethodInAnyScope

type DestructuringMethodInAnyScope
    = PreferDestructuringInExistingLet { direct : Bool, as_ : Bool }
    | UseDirectDestructuring { inExistingLet : Bool, as_ : Bool }
    | UseDestructuringAs { inExistingLet : Bool, direct : Bool }
```
or something similar... Thoughts?

**Comment by SiriusStarr** [comment]
Ahh, very good point about scope.  The most important thing for a fix would of course be not producing code that fails to compile (due to name clashes).

It's probably most clear to a user to think about name clashes as a fallback case; I would be surprised if separate `let`s would be the default anyone would want, since they're pretty rare to encounter in Elm code.

This is a first pass at options with documentation.  We allow the choice of where to destructure (in argument, which will use `as` if necessary, or in `let`), and then a choice of how to resolve name clashes (manually, which will not provide a fix, or in a separate `let`).

```elm
{-| Configure the rule, determining how automatic fixes are generated.
-}
type alias Config =
    ( DestructureIn, ResolveNameClash )


{-| Specify how any detected single-pattern cases should be fixed.

`DestructureInArgument` -- Fix by destructuring in the function argument. This will use an `as` pattern if the case expression is also used. For example:

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after (Opaque i) =
        i

and

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after ((Opaque i) as o) =
        f o i

`DestructureInLetBinding` -- Fix by destructuring in a `let` binding. For example:

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            (Opaque i) =
                o
        in
        i

and

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after o =
        let
            (Opaque i) =
                o
        in
        f o i

-}
type DestructureIn
    = DestructureInArgument
    | DestructureInLetBinding


{-| Fixing a single-pattern case expression by destructuring in either the
function argument or a `let` binding expands the scope of the variable. If
there is already a variable in scope with the same name, this would lead to a
name clash.

`Manually` -- Do not provide an automatic fix, allowing you to resolve the name clash manually.

`WithSeparateLet` -- Use a separate `let` binding to destructure, since the scope will then be different. For example:

    before : Opaque -> Int
    before o =
        let
            clash i =
                i
        in
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            clash i =
                i
        in
        let
            (Opaque i) =
                o
        in
        i

-}
type ResolveNameClash
    = Manually
    | WithSeparateLet
```

**Comment by lue-bird** [comment]
I generally really like this!
```elm
type ResolveNameClash
    = ResolveNameClashManually
    | ResolveNameClashWithSeparateLet
```
is 👍 and we could add some more options to
```elm
type DestructureIn =
    = DestructureInSeperateLet
    | DestructureInArgument { allowAs : Bool } ResolveNameClash
    | DestructureInExistingLets WhenNoLetsInScope ResolveNameClash

type WhenNoLetsInScope
    = WhenNoLetsInScopeTryDestructuringInArgument { allowAs : Bool }
    | WhenNoLetsInScopeUseSeparateLet

type alias Config =
    DestructureIn

myPersonalConfig =
    DestructureInExistingLets
      (WhenNoLetsInScopeTryDestructuringInArgument
          { allowAs = False }
      )
      ResolveNameClashWithSeparateLet
```
(probably needs better names)

**Comment by SiriusStarr** [comment]
Rather than trying to reason about the hierarchy of conditions, I'm thinking it might be more simple to just allow picking `let` or argument for each condition, like this:

```elm
{-| Configure the rule, determining how automatic fixes are generated.

  - `fixBy` -- Specify how to automatically fix single-pattern cases. `alwaysInArgument` and `alwaysInLetBinding` are available as helpful defaults, or you can specify more complex behavior.
  - `ifNameClash` -- Specify what to do if a pattern name clashes with a name already in scope.

-}
type alias Config =
    { fixBy : FixByDestructuringIn
    , ifNameClash : ResolveNameClash
    }


{-| Always fix by destructuring in the argument, even if an `as` pattern would be required.
-}
alwaysInArgument : FixByDestructuringIn
alwaysInArgument =
    FixByDestructuringIn Argument Argument Argument Argument


{-| Always fix by destructuring in a `let` binding, even it would have to be created.
-}
alwaysInLetBinding : FixByDestructuringIn
alwaysInLetBinding =
    FixByDestructuringIn LetBinding LetBinding LetBinding LetBinding


{-| Specify how to automatically fix single-pattern cases in specific conditions.


## `default`

Specify how to fix single-pattern cases that do not match a more complex condition.


### `default = Argument`

    before : Opaque -> Int
    before o =
        let
            foo =
                bar
        in
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after (Opaque i) =
        let
            foo =
                bar
        in
        i


### `default = LetBinding`

    before : Opaque -> Int
    before o =
        let
            foo =
                bar
        in
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            foo =
                bar

            (Opaque i) =
                o
        in
        i


## `asPatternRequired`

Specify how to fix single-pattern cases that would require the use of an `as` pattern if destructured in the argument


### `asPatternRequired = Argument`

    before : Opaque -> Int
    before o =
        let
            foo =
                bar
        in
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after ((Opaque i) as o) =
        let
            foo =
                bar
        in
        f o i


### `asPatternRequired = LetBinding`

    before : Opaque -> Int
    before o =
        let
            foo =
                bar
        in
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after o =
        let
            foo =
                bar

            (Opaque i) =
                o
        in
        f o i


## `noExistingLet`

Specify how to fix single-pattern cases when destructuring in a `let` would require the creation of one, since none already exists in scope.


### `noExistingLet = Argument`

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after (Opaque i) =
        i


### `noExistingLet = LetBinding`

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            (Opaque i) =
                o
        in
        i


## `noExistingLetAndAsPatternRequired`

Specify how to fix single-pattern cases when no `let` pattern already exists in scope and an `as` pattern would be required if destructured in the argument.


### `noExistingLetAndAsPatternRequired = Argument` -- Fix by destructuring in the function argument. For example:

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after ((Opaque i) as o) =
        f o i


### `noExistingLetAndAsPatternRequired = LetBinding`

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                f o i

    after : Opaque -> Int
    after o =
        let
            (Opaque i) =
                o
        in
        f o i

-}
type alias FixByDestructuringIn =
    { default : DestructureIn
    , asPatternRequired : DestructureIn
    , noExistingLet : DestructureIn
    , noExistingLetAndAsPatternRequired : DestructureIn
    }


{-| Specify how any detected single-pattern cases should be fixed.

`Argument` -- Fix by destructuring in the function argument. For example:

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after (Opaque i) =
        i

`LetBinding` -- Fix by destructuring in a `let` binding. For example:

    before : Opaque -> Int
    before o =
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            (Opaque i) =
                o
        in
        i

-}
type DestructureIn
    = Argument
    | LetBinding


{-| Fixing a single-pattern case expression by destructuring in either the
function argument or a `let` binding expands the scope of the variable. If
there is already a variable in scope with the same name, this would lead to a
name clash.

`Manually` -- Do not provide an automatic fix, allowing you to resolve the name clash manually.

`WithSeparateLet` -- Use a separate `let` binding to destructure, since the scope will then be different. For example:

    before : Opaque -> Int
    before o =
        let
            clash i =
                i
        in
        case o of
            Opaque i ->
                i

    after : Opaque -> Int
    after o =
        let
            clash i =
                i
        in
        let
            (Opaque i) =
                o
        in
        i

-}
type ResolveNameClash
    = Manually
    | WithSeparateLet
```

**Comment by lue-bird** [comment]
Should fix options like
- `{ default = LetBinding, asPatternRequired = Argument, ... }`
- `{ default = Argument, noExistingLet = LetBinding, ... }`

exist? Nobody would/should use these (if I understand correctly).

Sorting my thoughts: The valid `FixByDestructuring` options are

- Destructure in existing lets
    - | If no lets exist
        - Destructure the argument. If the var pattern is a record field pattern or `as` destructured or if between `case` and `of` is not just an argument, use _no other fix_
            - | If the argument is also used elsewhere
                - Destructure using `as`
                - _no other fix_
        - _no other fix_
- Destructure the argument
    - | If the argument is also used elsewhere
        - Destructure using `as`
        - Destructure in existing lets. 
            - | If no lets exist
                - Destructure using `as`
                - _no other fix_
    - | If the expression between `case` and `of` is not an argument or the argument pattern is a record field pattern or `as` destructured
        - Destructure in existing lets. If no lets exist, use _no other fix_
        - _no other fix_

When _no other fix_ (also for name clashes):
- Create separate let
- Don't provide a fix

Leaving out no options, we get

```elm
type alias Config =
    { fix : FixByDestructuringIn
    , noOtherOption : NoOtherOption
    }

{-|
- if names from the destructured pattern clash with existing argument names
- when choosing `NoOtherOptionFix`
-}
type NoOtherOption
    = CreateSeparateLet
    | NoFix

type OrNoOtherOptionFix option
    = Fix option
    | NoOtherOptionFix

type FixByDestructuringIn
    = FixByDestructuringInExistingLets
        { noExistingLets :
            OrNoOtherOptionFix
                { destructureTheArgument :
                    { argumentIsAlsoUsedElsewhere :
                        OrNoOtherOptionFix
                            DestructureUsingAs
                    -- If the var pattern is a record field pattern or `as` destructured, use _noOtherFix_
                    }
                }
        }
    | FixByDestructuringTheArgument
        { argumentIsAlsoUsedElsewhere :
            ArgumentIsAlsoUsedElsewhere
        , noArgumentInCaseOfOrArgumentNotDestructable :
          --| If between `case` and `of` isn't just an argument or
          --| record field pattern or `as` destructured
            OrNoOtherOptionFix
                DestructureInExistingLets -- If no lets exist, use _noOtherFix_
        }

type DestructureUsingAs = DestructureUsingAs
type DestructureInExistingLets = DestructureInExistingLets

type ArgumentIsAlsoUsedElsewhere
    = DestructureUsingAsInstead
    | DestructureInExistingLetsInstead
        { noExistingLetsWhereArgumentIsntUsed :
            OrNoOtherOptionFix
                DestructureUsingAs
        }
```

This _is_ more complex but it's a starting point.

>It might be more simple to just allow picking let or argument for each condition

That's an interesting idea for situations that might occur in all (sub-)options.

→ The hierarchy version might describe the desired fix better for the options I listed (?)

**Comment by SiriusStarr** [comment]
> Should fix options like
> 
>     * `{ default = LetBinding, asPatternRequired = Argument, ... }`
> 
>     * `{ default = Argument, noExistingLet = LetBinding, ... }`
> 
> 
> exist? Nobody would/should use these (if I understand correctly).

Yeah, I don't disagree that they'd be a weird choice, but I'm just trying to avoid a massive hierarchy of custom types that people have to try to wrap their minds around just for the happy path.  95% of the time, people are probably just going to want always argument or always `let`, so I'm not sure that letting people pick a weird set of choices if they choose to manually specify is a problem.  I think your types do a good job fully specifying all possible behavior, I just worry about the ergonomics of it.

Also, currently it's slightly non-intuitive that there isn't a "visible" option to just *create* a `let` if none exists.  That functionality is provided through `CreateSeperateLet`, but that might not be intuitive.

EDIT:  So maybe the best choice would be to start thinking about some sort of builder functions to hide this type complexity?

**Comment by lue-bird** [comment]
#### Status
- there are no real builders but an improved version similar to the union types
    - users don't choose `NoOtherOptionFix`. they either choose `createSeparateLet` or `noFix` which is checked to be equal to the ...onNameClash at compile time → more intuitive
    - all `Fix (DestructureSomewhereElse)` are created `destructureSomewhereElse`
    - default `Config`s `alwaysInLetBinding` and `alwaysInArgument`
- docs exist but could be improved with examples etc.
- tests cover most cases
- could get better names
    -  `notDestructurable`: for example `noArgumentInCaseOfOrArgumentNotDestructurable` but shorter?
    - `argumentAlsoUsedElsewhere`: `argumentUsedElsewhere`/...?
- so far the fixes it creates seem to work

Help (on the docs) appreciated 🙂

**Comment by SiriusStarr** [comment]
Don't worry about `elm-review` failures; I'll go through and clean those up before we merge, once we're satisfied with the rest.

Putting this here just to remind myself, but make sure we bump `elm-syntax-dsl` to `5.2.2` for the-sett/elm-syntax-dsl#31 (if there's any chance of writing a case expression out).

I've been very busy this week, apologies, so I'm not around much to help with stuff.  I will work on cleaning up the docs/examples when I have time.

Thanks for all of your work on this!

File: tests/NoSinglePatternCaseTest.elm
Comment (jfmengels) [inline tests/NoSinglePatternCaseTest.elm]: Just want to point out that [`Simplify`](https://package.elm-lang.org/packages/jfmengels/elm-review-simplify/latest/Simplify) already makes this change if the pattern does not introduce any variable.

**Comment by SiriusStarr** [comment]
This should be about ready for merging now, if you want to give it a once-over.  :+1:

**Review by SiriusStarr — APPROVED** [review]

**Pull request merged**
